packages/core/src/lib/lit-core.ts

```diff
@@ -330,39 +328,55 @@ export class LitCore {
-                Such that the state when the request was sent will now mutate when the response is sent back.
-                The sdk should be able to understand its current execution environment and wait on an active
-                network request to the previous epoch's node set before changing over.
-              */
+                  TODO: This covers *most* cases where a node may come in or out of the active
+                  set which we will need to re attest to the execution environments.
+                  However, the sdk currently does not know if there is an active network operation pending.
+                  Such that the state when the request was sent will now mutate when the response is sent back.
+                  The sdk should be able to understand its current execution environment and wait on an active
+                  network request to the previous epoch's node set before changing over.
+                */
             log(
               'Active validator sets changed, new validators ',
               delta,
               'starting node connection'
             );
+            await this.connect(); // Will update `epochInfo`
           }
-
-          await this.connect();
-        } catch (err: unknown) {
-          // FIXME: We should emit an error event so that consumers know that we are de-synced and can connect() again
-          // But for now, our every-30-second network sync will fix things in at most 30s from now.
-          // this.ready = false; Should we assume core is invalid if we encountered errors refreshing from an epoch change?
-          const { message = '' } = err as
-            | Error
-            | NodeClientErrorV0
-            | NodeClientErrorV1;
-          logError(
-            'Error while attempting to reconnect to nodes after epoch transition:',
-            message
+        } else {
+          // In case of centralised networks, we don't run `connect()` flow, so we will manually update epochInfo here
+          this._epochState = await this._fetchCurrentEpochState(
+            validatorData.epochInfo
           );
         }
       }
+    } catch (err: unknown) {
+      // Ensure that any methods that check `this.ready` throw errors to the caller, and any consumers can check appropriately
+      this.ready = false;
+
+      const { message = '' } = err as
+        | Error
+        | NodeClientErrorV0
+        | NodeClientErrorV1;
+      logError(
+        'Error while attempting to reconnect to nodes after epoch transition:',
+        message
+      );
+
+      // Signal to any listeners that we've encountered a fatal error
+      this.emit(
+        'error',
+        new Error(
+          'Error while attempting to reconnect to nodes after epoch transition:' +
+            message
+        )
+      );
+
+      // Signal to any listeners that we're 'disconnected' from LIT network
+      this.emit('disconnected', true);
```

Comment:
I would prefer listening to only one event. We have `connected` below. Here we can emit:
```suggestion
      this.emit('connected', false);
```

As a listener I would expect to listen to connection status as this is likely the status I'm interested in (to know if I can use Lit), instead of just one of those events

Reply:
Emitting `disconnect` and `connect` events is a pretty stand/common pattern, since the action taken is typically very different for each case; it's very common to define `onDisconnect` and `onConnect` functions in the app logic layer and just bind them directly as event handlers. If we emit a single `connection` event with the state of the connection, then every event handler in the app layer is probably going to start with an if() else {} block that needs to check the payload emitted to the event, and identify the type, and call the appropriate connect or disconnect handler.

Also, most consumers will actually not really care about _connected_ events because the `connect()` method itself returns a promise that they can simply wait on -- but listening to `disconnect` events is actually quite important for all consumers.

I personally think emitting separate events is a nice devex -- but I've got no argument against adding another _connection_ type event that is emitted every time the connection state changes, either connecting or disconnecting, if you think that would be a good addition to the API :+1:

One thing to consider is that we can also strongly type `disconnectHandler` and `connectHandler` functions -- and the payload provided to the listener would likely make sense to actually be a different type.  Actually, I think that my making this always emitt a boolean is probably a poor API, now that you mention it -- what do you think about modifying the `disconnect` event to emit the `Error` that was encountered during the disconnect, e.g. `this.emit('disconnected', { reason: error })`?

Reply:
Great points. The thing that tickled me was that we were emitting events with a boolean that is always the same, seems like it is neither a `connected` status API nor a `onConnect` + `onDisconnect` events

But your last suggestion I think nails the thing. And we can make something similar for `connected` when we see the need for it